shared/src/main/java/org/geysermc/hydraulic/pack/PackManager.java

```diff
@@ -157,4 +208,108 @@ private void callEvent(@NotNull ModInfo mod, @NotNull Event event) {
             module.call(event.getClass(), new PackEventContext(this.hydraulic, mod, module, event));
         }
     }
+
+    private void initializeModLookups() {
+        // Step 1: Lookup which namespaces are contained by which mods
+        final Multimap<String, ModInfo> namespacesToMods = this.namespacesToMods;
+        namespacesToMods.clear();
+        for (final ModInfo mod : hydraulic.mods()) {
+            for (final Path root : mod.roots()) {
+                final Path assets = root.resolve("assets");
+                if (!Files.isDirectory(assets)) continue;
+                try (Stream<Path> stream = Files.list(assets)) {
+                    stream.filter(Files::isDirectory)
+                        .map(Path::getFileName)
+                        .map(Path::toString)
+                        .filter(namespace -> !namespace.equals("minecraft"))
+                        .forEach(namespace -> namespacesToMods.put(namespace, mod));
+                } catch (IOException e) {
+                    LOGGER.error("Failed to list namespaces for mod {}", mod.id(), e);
+                }
+            }
+        }
+
+        // Step 2: Use namespace information to lookup which mods contains what block models
+        final Multimap<String, ResourceLocation> modsToBlocks = this.modsToBlocks;
+        modsToBlocks.clear();
+        blocksLoop:
+        for (final ResourceLocation block : BuiltInRegistries.BLOCK.keySet()) {
+            if (block.getNamespace().equals("minecraft")) continue;
+            for (final ModInfo mod : namespacesToMods.get(block.getNamespace())) {
+                final Path checkFile = mod.resolveFile("assets/" + block.getNamespace() + "/blockstates/" + block.getPath() + ".json");
+                if (checkFile != null) {
+                    modsToBlocks.put(mod.id(), block);
+                    continue blocksLoop;
```

Comment:
This can just be a break as that will break the inner loop and return to the outer